opensearch-project/OpenSearch-Dashboards
# [Workspace] Add workspaceAvailability field into application

### Description


Home page, as well as workspace create page / workspace list page, which is supposed to be only visited out of any workspace, now is enabled to be accessed by clicking the home icon or enter the url manually in the browser.

This PR introduces a new field `workspaceAvailability` inside App, which indicates the availability of the application based on workspace. This field can bring two benefits:
1. When navigate between apps, plugins and core application are using the `navigateToApp` method. By using the workspaceAvailability field, workspace is able to do the hard navigation in a central place.
2. The workspaceAvailability field can be used in workspace create page to filter applications that should not be visible inside workspace. For now we filter out home overview page and all the pages under management section by hard code, which is not elegant and not extensible.

### Issues Resolved



closes #6362 

## Screenshot


### 1. Clicking home icon will exist workspace.
![20240407094009141](https://github.com/ruanyl/OpenSearch-Dashboards/assets/13493605/f6937943-2d95-4bfb-b8ca-27631ce436a4)

### 2. Apps with `workspaceAvailability.outOfWorkspace` is inaccessible within workspace
![image](https://github.com/opensearch-project/OpenSearch-Dashboards/assets/13493605/3db34b85-f721-4544-960f-b9e0ddaa9a34)

## Testing the changes


- Using the branch to bootstrap
- enabled `workspace.enabled` to true
- Start OSD by using `yarn start --no-base-path` to make sure no random base path will be appended
- Navigate to devTools
- Insert test workspaces by calling
```
PUT .kibana/_doc/workspace:foo
{
  "type": "workspace",
  "workspace": {
    "name": "foo"
  }
}
```
- Visit the homepage inside workspace: `http://localhost:5601/w/foo/app/home`
- You will find an error page will be found saying the application can not be found.
- Go into a page within workspace `http://localhost:5601/w/foo/app/workspace_update`
- Clicking the home icon in the top navigation
- You will you being navigated to home page out of workspace.

## Changelog

- feat: [Workspace] Allow making apps available in workspaces using `workspaceAvailability`

### Check List
